### PR TITLE
core/atomic.d: Add GDC atomic support code

### PR DESCRIPTION
**Pull request by ibuclaw**
Following on from #2596, rewrite to work with the new atomic API.

This has been peer reviewed by @TurkeyMan 

Best viewed with whitespace ignored, as the removals are only highlighted by indenting the DMD specific code into a version(DigitalMars) block.

### CODE REVIEW

File: src/core/internal/atomic.d [hunk 1]
```diff
@@ -12,664 +12,1075 @@ module core.internal.atomic;
...
+    }
 
-            Lend:
+    private bool atomicCompareExchangeImpl(MemoryOrder succ = MemoryOrder.seq, MemoryOrder fail = MemoryOrder.seq, bool weak, T)(T* dest, T* compare, T value) pure nothrow @nogc @trusted
+        if (CanCAS!T)
+    {
+        bool res = void;
 
-                ret;
+        static if (GNU_Have_Atomics || GNU_Have_LibAtomic)
```
Comment (TurkeyMan) [inline src/core/internal/atomic.d]: 3 out if 5 lines in this section is wasted vertical space :/
Most of the lines here are `{`, `}` and the line split, which is split well before the horiz white space even github shows here in the review render ;)

The 3 if/else blocks up a little waste 50% if lines to `{`, `}`, but this blocks triggers me ;)
Where I'm from, monitors have been 16/9 for almost at least 15 years... so I feel like vertical space is about 16/9th's more valuable ;)

**Comment by TurkeyMan** [comment]
What's sad is that no unit tests will cover this GDC code... is that fine?

File: src/core/internal/atomic.d (hunk 1 above)
Comment (wilzbach) [reply src/core/internal/atomic.d]: https://dlang.org/dstyle

Comment (ibuclaw) [reply src/core/internal/atomic.d]: I probably have a subconscious habit of keeping to 80 chars as well.

**Comment by ibuclaw** [comment]
> What's sad is that no unit tests will cover this GDC code... is that fine?

Erm, all the unittests in core.atomic.

File: src/core/internal/atomic.d (hunk 1 above)
Comment (ibuclaw) [reply src/core/internal/atomic.d]: Any though it's just an opinion, I don't personally find compressing the format very readable.

image

Comment (ibuclaw) [reply src/core/internal/atomic.d]: Removed some braces just to make @TurkeyMan happy.

**Comment by TurkeyMan** [comment]
Yeah but the DMD CI won't prove this code path, it could regress against GDC?

File: src/core/internal/atomic.d (hunk 1 above)
Comment (TurkeyMan) [reply src/core/internal/atomic.d]: left the awkward line breaks though :P

Comment (ibuclaw) [reply src/core/internal/atomic.d]: It's there for readability. :-P

**Comment by ibuclaw** [comment]
> Yeah but the DMD CI won't prove this code path, it could regress against GDC?

If it isn't syntactically valid, an error would be thrown.  Any API changes need to be synchronized, however I would expect the person making the change would do the required due dilligence, otherwise there's no basis for accepting them.

**Comment by kinke** [comment]
Iain, what's your ultimate goal wrt. these upstreamings? Do you aim for not having to maintain druntime/Phobos forks or do you plan to restrict the upstreamings to manageably-sized diffs?

**Pull request merged**